pkg/agent/wireguard/client_linux.go

```diff
@@ -0,0 +1,382 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wireguard
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"sync"
+
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"golang.zx2c4.com/wireguard/wgctrl"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/types"
+	"antrea.io/antrea/pkg/agent/util/sysctl"
+)
+
+const (
+	ifname            = "antrea-wg"
+	listenPort        = 51850
+	routeMask         = 0x100
+	routeMark         = 0x100
+	routeRulePriority = 10
+	routeTable        = 10
+)
+
+var zeroKey = wgtypes.Key{}
+
+// wireguardClient is an interface to mock wgctrl.Client
+type wireguardClient interface {
+	io.Closer
+	Devices() ([]*wgtypes.Device, error)
+	Device(name string) (*wgtypes.Device, error)
+	ConfigureDevice(name string, config wgtypes.Config) error
+}
+
+var _ Interface = (*client)(nil)
+
+type client struct {
+	sync.Mutex
+	wgClient                 wireguardClient
+	nodeName                 string
+	k8sclient                clientset.Interface
+	mtu                      int
+	listenPort               int
+	privateKey               wgtypes.Key
+	peerByNodeName           map[string]wgtypes.PeerConfig
+	enabledIPv4, enabledIPv6 bool
+	nodeIPv4, nodeIPv6       net.IP
+}
+
+func New(nodeName string, k8sclient clientset.Interface, mtu int, nodeIPv4, nodeIPv6 net.IP, enabledIPv4, enabledIPv6 bool, port int) (Interface, error) {
+	if port == 0 {
+		port = listenPort
+	}
+	wgClient, err := wgctrl.New()
+	if err != nil {
+		return nil, err
+	}
+	return &client{
+		wgClient:       wgClient,
+		nodeName:       nodeName,
+		k8sclient:      k8sclient,
+		mtu:            mtu,
+		listenPort:     port,
+		nodeIPv4:       nodeIPv4,
+		nodeIPv6:       nodeIPv6,
+		enabledIPv4:    enabledIPv4,
+		enabledIPv6:    enabledIPv6,
+		peerByNodeName: make(map[string]wgtypes.PeerConfig),
+	}, nil
+}
+
+func (client *client) Init() error {
+	link := &netlink.Wireguard{LinkAttrs: netlink.LinkAttrs{Name: ifname}}
+	err := netlink.LinkAdd(link)
+	// ignore existing link as it might be managed by userspace process
+	if err != nil && !errors.Is(err, unix.EEXIST) {
+		if errors.Is(err, unix.EOPNOTSUPP) {
+			return fmt.Errorf("wireguard not supported by the Linux kernel (netlink: %w)", err)
+		}
+		return err
+	}
+	if err := netlink.LinkSetUp(link); err != nil {
+		return err
+	}
+	if err := netlink.LinkSetMTU(link, client.mtu); err != nil {
+		return err
+	}
+	if client.enabledIPv4 {
+		if sysctl.EnsureSysctlNetValue("ipv4/conf/all/rp_filter", 0) != nil {
+			return fmt.Errorf("setting net.ipv4.conf.all.rp_filter failed: %w", err)
```

Comment:
I am not too familiar with the design any more, but
1) if your solution avoids the extra encap overhead (Geneve + WG) then it's a good thing
2) I want to avoid adding any more calls to `EnsureSysctlNetValue` as this is just a step backwards in terms of supporting running the agent with privileged mode, which can be an issue for some K8s distributions

Reply:
The idea seems like we can replace the Geneve encapsulation with WireGuard if available. Then we can get rid of the policy routing and the change of rp_filter. I think it is doable but I am not sure whether this will introduce any side effects or not.

Reply:
Add please add comments to explain these operations if we need them.